RewriteTest validation fails on a DeclarativeRecipe calling an imperative recipe with no arguments

## What version of OpenRewrite are you using?
I am using
- OpenRewrite v8.1.8

## What is the smallest, simplest way to reproduce the problem?
```java
import org.junit.jupiter.api.Test;
import org.openrewrite.config.Environment;
import org.openrewrite.config.YamlResourceLoader;
import org.openrewrite.properties.Assertions;
import org.openrewrite.test.RewriteTest;

import java.io.ByteArrayInputStream;
import java.net.URI;
import java.util.Properties;

public class BugTest implements RewriteTest {
    @Test   // fails!
    void instantiateYamlRecipeReferencingImperativeRecipeWithoutArguments() {
        rewriteRun(
                spec -> spec.recipe(new ByteArrayInputStream(
                        """
                        ---
                        type: specs.openrewrite.org/v1beta/recipe
                        name: test.recipe
                        displayName: Test Recipe
                        recipeList:
                          - org.openrewrite.maven.RemoveRedundantDependencyVersions
                        """
                .getBytes()), "test.recipe"),
                Assertions.properties(""));
    }

    @Test    // success
    void instantiateYamlRecipeReferencingImperativeRecipeWithArguments() {
        rewriteRun(
                spec -> spec.recipe(new ByteArrayInputStream(
                        """
                        ---
                        type: specs.openrewrite.org/v1beta/recipe
                        name: test.recipe
                        displayName: Test Recipe
                        recipeList:
                          - org.openrewrite.maven.RemoveRedundantDependencyVersions:
                              junk: junk
                        """
                                .getBytes()), "test.recipe"),
                Assertions.properties(""));
    }

    @Test    // success
    void workaround() {
        rewriteRun(
                spec -> spec.recipe(Environment.builder()
                        .load(new YamlResourceLoader(new ByteArrayInputStream(
                                """
                                ---
                                type: specs.openrewrite.org/v1beta/recipe
                                name: org.openrewrite.maven.RemoveRedundantDependencyVersions
                                displayName: workaround
                                recipeList:
                                  - org.openrewrite.maven.RemoveRedundantDependencyVersions:
                                      junk: junk
                                """
                                        .getBytes()), URI.create("rewrite.yml"), new Properties()))
                        .load(new YamlResourceLoader(new ByteArrayInputStream(
                                """
                                ---
                                type: specs.openrewrite.org/v1beta/recipe
                                name: test.recipe
                                displayName: Test Recipe
                                recipeList:
                                  - org.openrewrite.maven.RemoveRedundantDependencyVersions
                                """
                                        .getBytes()), URI.create("rewrite.yml"), new Properties()))
                        .build()
                        .activateRecipes("test.recipe")),
                Assertions.properties(""));
    }
}

```

## What did you expect to see?
All success

## What did you see instead?
The first test case fails

## What is the full stack trace of any errors you encountered?
```
java.lang.AssertionError: [Recipe validation must have no failures] 
Expecting no elements of:
  [[Valid{property='initialization', value='org.openrewrite.config.DeclarativeRecipe@cf05ce89'},
    Invalid{property='test.recipe.recipeList[0] (in rewrite.yml)', value='org.openrewrite.maven.RemoveRedundantDependencyVersions', message='recipe 'org.openrewrite.maven.RemoveRedundantDependencyVersions' does not exist.'},
    Valid{property='initialization', value='org.openrewrite.config.DeclarativeRecipe@cf05ce89'},
    Invalid{property='test.recipe.recipeList[0] (in rewrite.yml)', value='org.openrewrite.maven.RemoveRedundantDependencyVersions', message='recipe 'org.openrewrite.maven.RemoveRedundantDependencyVersions' does not exist.'}],
    [Valid{property='initialization', value='org.openrewrite.config.DeclarativeRecipe@cf05ce89'},
    Invalid{property='test.recipe.recipeList[0] (in rewrite.yml)', value='org.openrewrite.maven.RemoveRedundantDependencyVersions', message='recipe 'org.openrewrite.maven.RemoveRedundantDependencyVersions' does not exist.'}]]
to match given predicate but this element did:
  [Valid{property='initialization', value='org.openrewrite.config.DeclarativeRecipe@cf05ce89'},
    Invalid{property='test.recipe.recipeList[0] (in rewrite.yml)', value='org.openrewrite.maven.RemoveRedundantDependencyVersions', message='recipe 'org.openrewrite.maven.RemoveRedundantDependencyVersions' does not exist.'},
    Valid{property='initialization', value='org.openrewrite.config.DeclarativeRecipe@cf05ce89'},
    Invalid{property='test.recipe.recipeList[0] (in rewrite.yml)', value='org.openrewrite.maven.RemoveRedundantDependencyVersions', message='recipe 'org.openrewrite.maven.RemoveRedundantDependencyVersions' does not exist.'}]
```

## Are you interested in [contributing a fix to OpenRewrite](https://github.com/openrewrite/.github/blob/main/CONTRIBUTING.md#contributing-fixes)?
I think for this one, y'all would fix it better/faster

--- a/rewrite-core/src/main/java/org/openrewrite/config/ClasspathScanningLoader.java
+++ b/rewrite-core/src/main/java/org/openrewrite/config/ClasspathScanningLoader.java
@@ -99,6 +99,16 @@ public class ClasspathScanningLoader implements ResourceLoader {
                 .acceptPaths("META-INF/rewrite"), properties, dependencyResourceLoaders, classLoader);
     }
 
+    public static ClasspathScanningLoader onlyYaml(Properties properties) {
+        ClasspathScanningLoader classpathScanningLoader = new ClasspathScanningLoader();
+        classpathScanningLoader.scanYaml(new ClassGraph().acceptPaths("META-INF/rewrite"),
+                properties, emptyList(), null);
+        return classpathScanningLoader;
+    }
+
+    private ClasspathScanningLoader() {
+    }
+
     /**
      * This must be called _after_ scanClasses or the descriptors of declarative recipes will be missing any
      * non-declarative recipes they depend on that would be discovered by scanClasses

--- a/rewrite-core/src/main/java/org/openrewrite/config/DeclarativeRecipe.java
+++ b/rewrite-core/src/main/java/org/openrewrite/config/DeclarativeRecipe.java
@@ -120,14 +120,8 @@ public class DeclarativeRecipe extends Recipe {
         uninitializedRecipes.add(recipe);
     }
 
-    public void addUninitialized(String recipeName, @Nullable ClassLoader classLoader) {
-        try {
-            uninitializedRecipes.add((Recipe) Class.forName(recipeName, true, classLoader != null ? classLoader : this.getClass().getClassLoader())
-                    .getDeclaredConstructor()
-                    .newInstance());
-        } catch (Exception e) {
-            uninitializedRecipes.add(new DeclarativeRecipe.LazyLoadedRecipe(recipeName));
-        }
+    public void addUninitialized(String recipeName) {
+        uninitializedRecipes.add(new DeclarativeRecipe.LazyLoadedRecipe(recipeName));
     }
 
     public void addValidation(Validated<Object> validated) {

--- a/rewrite-core/src/main/java/org/openrewrite/config/Environment.java
+++ b/rewrite-core/src/main/java/org/openrewrite/config/Environment.java
@@ -229,6 +229,10 @@ public class Environment {
             return load(new ClasspathScanningLoader(properties, classLoader));
         }
 
+        public Builder scanYamlResources() {
+            return load(ClasspathScanningLoader.onlyYaml(properties));
+        }
+
         /**
          * @param jar         A path to a jar file to scan.
          * @param classLoader A classloader that is populated with the transitive dependencies of the jar.

--- a/rewrite-core/src/main/java/org/openrewrite/config/YamlResourceLoader.java
+++ b/rewrite-core/src/main/java/org/openrewrite/config/YamlResourceLoader.java
@@ -260,15 +260,29 @@ public class YamlResourceLoader implements ResourceLoader {
                             int i,
                             Object recipeData) {
         if (recipeData instanceof String) {
-            recipe.addUninitialized((String) recipeData, classLoader);
+            String recipeName = (String) recipeData;
+            try {
+                // first try an explicitly-declared zero-arg constructor
+                recipe.addUninitialized((Recipe) Class.forName(recipeName, true,
+                                classLoader == null ? this.getClass().getClassLoader() : classLoader)
+                        .getDeclaredConstructor()
+                        .newInstance());
+            } catch (ReflectiveOperationException reflectiveOperationException) {
+                try {
+                    // then try jackson
+                    recipe.addUninitialized(instantiateRecipe(recipeName, new HashMap<>()));
+                } catch (IllegalArgumentException illegalArgumentException) {
+                    // else, it's probably declarative
+                    recipe.addUninitialized(recipeName);
+                }
+            }
         } else if (recipeData instanceof Map) {
             Map.Entry<String, Object> nameAndConfig = ((Map<String, Object>) recipeData).entrySet().iterator().next();
             try {
                 if (nameAndConfig.getValue() instanceof Map) {
-                    Map<Object, Object> withJsonType = new HashMap<>((Map<String, Object>) nameAndConfig.getValue());
-                    withJsonType.put("@c", nameAndConfig.getKey());
                     try {
-                        recipe.addUninitialized(mapper.convertValue(withJsonType, Recipe.class));
+                        recipe.addUninitialized(instantiateRecipe(nameAndConfig.getKey(),
+                                (Map<String, Object>) nameAndConfig.getValue()));
                     } catch (IllegalArgumentException e) {
                         if (e.getCause() instanceof InvalidTypeIdException) {
                             recipe.addValidation(Validated.invalid(nameAndConfig.getKey(),
@@ -297,6 +311,12 @@ public class YamlResourceLoader implements ResourceLoader {
                     "is an object type that isn't recognized as a recipe.",
                     null));
         }
+    }
+
+    private Recipe instantiateRecipe(String recipeName, Map<String, Object> args) throws IllegalArgumentException {
+        Map<Object, Object> withJsonType = new HashMap<>(args);
+        withJsonType.put("@c", recipeName);
+        return mapper.convertValue(withJsonType, Recipe.class);
     }
 
     @Override

--- a/rewrite-core/src/main/java/org/openrewrite/internal/lang/NullUtils.java
+++ b/rewrite-core/src/main/java/org/openrewrite/internal/lang/NullUtils.java
@@ -95,12 +95,20 @@ public class NullUtils {
         for (Field field : fields) {
             field.setAccessible(true);
             if(fieldHasNonNullableAnnotation(field) ||
-                    (defaultNonNull && !fieldHasNullableAnnotation(field))) {
+                    (defaultNonNull && !fieldHasNullableAnnotation(field)) ||
+                    fieldIsRequiredOption(field)) {
                 nonNullFields.add(field);
             }
         }
         nonNullFields.sort(Comparator.comparing(Field::getName));
         return nonNullFields;
+    }
+
+    private static boolean fieldIsRequiredOption(Field field) {
+        Option annotation = field.getAnnotation(Option.class);
+        if (annotation != null)
+            return annotation.required();
+        return false;
     }
 
     private static boolean fieldHasNonNullableAnnotation(Field field) {

--- a/rewrite-core/src/main/java/org/openrewrite/internal/lang/NullUtils.java
+++ b/rewrite-core/src/main/java/org/openrewrite/internal/lang/NullUtils.java
@@ -15,6 +15,9 @@
  */
 package org.openrewrite.internal.lang;
 
+import org.openrewrite.Option;
+
+import java.lang.annotation.Annotation;
 import java.lang.reflect.Field;
 import java.util.*;
 
@@ -48,7 +51,7 @@ public class NullUtils {
      * <li>org.checkerframework.checker.nullness.qual.NonNull</li>
      * <li>javax.validation.constraints.NotNull</li>
      */
-    private static List<String> FIELD_LEVEL_NON_NULL_ANNOTATIONS = Arrays.asList(
+    private static final List<String> FIELD_LEVEL_NON_NULL_ANNOTATIONS = Arrays.asList(
             "NonNull",
             "Nonnull",
             "NotNull"
@@ -67,7 +70,7 @@ public class NullUtils {
      * <li>org.checkerframework.checker.nullness.qual.Nullable</li>
      * <li>javax.validation.constraints.NotNull</li>
      */
-    private static List<String> FIELD_LEVEL_NULLABLE_ANNOTATIONS = Collections.singletonList(
+    private static final List<String> FIELD_LEVEL_NULLABLE_ANNOTATIONS = Collections.singletonList(
             "Nullable"
     );
 
@@ -112,13 +115,21 @@ public class NullUtils {
     }
 
     private static boolean fieldHasNonNullableAnnotation(Field field) {
-        return Arrays.stream(field.getDeclaredAnnotations())
-                .map(a -> a.annotationType().getSimpleName())
-                .anyMatch(FIELD_LEVEL_NON_NULL_ANNOTATIONS::contains);
+        for (Annotation a : field.getDeclaredAnnotations()) {
+            String simpleName = a.annotationType().getSimpleName();
+            if (FIELD_LEVEL_NON_NULL_ANNOTATIONS.contains(simpleName)) {
+                return true;
+            }
+        }
+        return false;
     }
     private static boolean fieldHasNullableAnnotation(Field field) {
-        return Arrays.stream(field.getDeclaredAnnotations())
-                .map(a -> a.annotationType().getSimpleName())
-                .anyMatch(FIELD_LEVEL_NULLABLE_ANNOTATIONS::contains);
+        for (Annotation a : field.getDeclaredAnnotations()) {
+            String simpleName = a.annotationType().getSimpleName();
+            if (FIELD_LEVEL_NULLABLE_ANNOTATIONS.contains(simpleName)) {
+                return true;
+            }
+        }
+        return false;
     }
 }

--- a/rewrite-core/src/main/java/org/openrewrite/internal/lang/NullUtils.java
+++ b/rewrite-core/src/main/java/org/openrewrite/internal/lang/NullUtils.java
@@ -76,8 +76,11 @@ public class NullUtils {
 
     /**
      * The method uses reflection to find all declared fields of a class that have been marked (via commonly used
-     * annotations) as being Non-Null. This method will also look at the class's package level to see if the API
-     * for that package is defaulted as Non-Null. ANY annotation that has runtime retention and matches on of the simple
+     * annotations) as being Non-Null, or a required {@link Option}.
+     * This method will also look at the class's package level to see if the API
+     * for that package is defaulted as Non-Null.
+     * Fields with explicit Nullable annotations will be excluded.
+     * Any other annotation that has runtime retention and matches one of the simple
      * annotation names (minus any package) will be considered a match.
      *
      * @param _class The class to reflect over
@@ -97,9 +100,10 @@ public class NullUtils {
         List<Field> nonNullFields = new ArrayList<>(fields.length);
         for (Field field : fields) {
             field.setAccessible(true);
-            if(fieldHasNonNullableAnnotation(field) ||
-                    (defaultNonNull && !fieldHasNullableAnnotation(field)) ||
-                    fieldIsRequiredOption(field)) {
+            if (fieldHasNullableAnnotation(field)) {
+                continue;
+            }
+            if (defaultNonNull || fieldHasNonNullableAnnotation(field) || fieldIsRequiredOption(field)) {
                 nonNullFields.add(field);
             }
         }

--- a/rewrite-core/src/main/java/org/openrewrite/text/FindAndReplace.java
+++ b/rewrite-core/src/main/java/org/openrewrite/text/FindAndReplace.java
@@ -79,6 +79,7 @@ public class FindAndReplace extends Recipe {
                           "Multiple patterns may be specified, separated by a semicolon `;`. " +
                           "If multiple patterns are supplied any of the patterns matching will be interpreted as a match. " +
                           "When not set, all source files are searched. ",
+            required = false,
             example = "**/*.java")
     @Nullable
     String filePattern;

--- a/rewrite-core/src/test/java/org/openrewrite/RecipeLifecycleTest.java
+++ b/rewrite-core/src/test/java/org/openrewrite/RecipeLifecycleTest.java
@@ -15,19 +15,25 @@
  */
 package org.openrewrite;
 
+import lombok.NoArgsConstructor;
 import org.intellij.lang.annotations.Language;
 import org.junit.jupiter.api.Test;
+import org.openrewrite.config.Environment;
 import org.openrewrite.config.RecipeDescriptor;
+import org.openrewrite.config.YamlResourceLoader;
 import org.openrewrite.internal.lang.Nullable;
 import org.openrewrite.marker.Markers;
 import org.openrewrite.test.RewriteTest;
 import org.openrewrite.text.PlainText;
 import org.openrewrite.text.PlainTextVisitor;
 
+import java.io.ByteArrayInputStream;
+import java.net.URI;
 import java.nio.charset.Charset;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.List;
+import java.util.Properties;
 import java.util.UUID;
 
 import static org.assertj.core.api.Assertions.assertThat;
@@ -248,5 +254,164 @@ class RecipeLifecycleTest implements RewriteTest {
                 return super.visitText(text, executionContext);
             }
         }).withName(name);
+    }
+
+    @Test
+    void canCallImperativeRecipeWithoutArgsFromDeclarative() {
+        rewriteRun(spec -> spec.recipeFromYaml("""
+            ---
+            type: specs.openrewrite.org/v1beta/recipe
+            name: test.recipe
+            displayName: Test Recipe
+            recipeList:
+              - org.openrewrite.NoArgRecipe
+            """,
+          "test.recipe"
+          ),
+          text("Hi", "NoArgRecipeHi"));
+    }
+
+    @Test
+    void canCallImperativeRecipeWithUnnecessaryArgsFromDeclarative() {
+        rewriteRun(spec -> spec.recipeFromYaml("""
+            ---
+            type: specs.openrewrite.org/v1beta/recipe
+            name: test.recipe
+            displayName: Test Recipe
+            recipeList:
+              - org.openrewrite.NoArgRecipe:
+                  foo: bar
+            """,
+            "test.recipe"
+          ),
+          text("Hi", "NoArgRecipeHi"));
+    }
+
+    @Test
+    void canCallRecipeWithNoExplicitConstructor() {
+        rewriteRun(spec -> spec.recipeFromYaml("""
+            ---
+            type: specs.openrewrite.org/v1beta/recipe
+            name: test.recipe
+            displayName: Test Recipe
+            recipeList:
+              - org.openrewrite.DefaultConstructorRecipe
+            """,
+            "test.recipe"
+          ),
+          text("Hi", "DefaultConstructorRecipeHi"));
+    }
+
+    @Test
+    void declarativeRecipeChain() {
+        rewriteRun(spec -> spec.recipeFromYaml("""
+            ---
+            type: specs.openrewrite.org/v1beta/recipe
+            name: test.recipe.a
+            displayName: Test Recipe
+            recipeList:
+              - test.recipe.b
+            ---
+            type: specs.openrewrite.org/v1beta/recipe
+            name: test.recipe.b
+            displayName: Test Recipe
+            recipeList:
+              - test.recipe.c
+            ---
+            type: specs.openrewrite.org/v1beta/recipe
+            name: test.recipe.c
+            displayName: Test Recipe
+            recipeList:
+              - org.openrewrite.NoArgRecipe
+            """,
+            "test.recipe.a"
+          ),
+          text("Hi", "NoArgRecipeHi"));
+    }
+
+    @Test
+    void declarativeRecipeChainAcrossFiles() {
+        rewriteRun(spec -> spec.recipe(Environment.builder()
+            .load(new YamlResourceLoader(new ByteArrayInputStream("""
+                ---
+                type: specs.openrewrite.org/v1beta/recipe
+                name: test.recipe.c
+                displayName: Test Recipe
+                recipeList:
+                  - org.openrewrite.NoArgRecipe
+                """.getBytes()),
+              URI.create("rewrite.yml"), new Properties()))
+            .load(new YamlResourceLoader(new ByteArrayInputStream("""
+                ---
+                type: specs.openrewrite.org/v1beta/recipe
+                name: test.recipe.b
+                displayName: Test Recipe
+                recipeList:
+                  - test.recipe.c
+                """.getBytes()),
+              URI.create("rewrite.yml"), new Properties()))
+            .load(new YamlResourceLoader(new ByteArrayInputStream("""
+                ---
+                type: specs.openrewrite.org/v1beta/recipe
+                name: test.recipe.a
+                displayName: Test Recipe
+                recipeList:
+                  - test.recipe.b
+                """.getBytes()),
+              URI.create("rewrite.yml"), new Properties()))
+            .build()
+            .activateRecipes("test.recipe.a")),
+          text("Hi", "NoArgRecipeHi"));
+    }
+}
+
+class DefaultConstructorRecipe extends Recipe {
+    @Override
+    public String getDisplayName() {
+        return "DefaultConstructorRecipe";
+    }
+
+    @Override
+    public String getDescription() {
+        return "DefaultConstructorRecipe.";
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new PlainTextVisitor<>() {
+            @Override
+            public PlainText visitText(PlainText text, ExecutionContext executionContext) {
+                if (!text.getText().contains(getDisplayName())) {
+                    return text.withText(getDisplayName() + text.getText());
+                }
+                return super.visitText(text, executionContext);
+            }
+        };
+    }
+}
+
+@NoArgsConstructor
+class NoArgRecipe extends Recipe {
+    @Override
+    public String getDisplayName() {
+        return "NoArgRecipe";
+    }
+
+    @Override
+    public String getDescription() {
+        return "NoArgRecipe.";
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new PlainTextVisitor<>() {
+            @Override
+            public PlainText visitText(PlainText text, ExecutionContext executionContext) {
+                if (!text.getText().contains(getDisplayName())) {
+                    return text.withText(getDisplayName() + text.getText());
+                }
+                return super.visitText(text, executionContext);
+            }
+        };
     }
 }

--- a/rewrite-core/src/test/java/org/openrewrite/RecipeLifecycleTest.java
+++ b/rewrite-core/src/test/java/org/openrewrite/RecipeLifecycleTest.java
@@ -363,6 +363,31 @@ class RecipeLifecycleTest implements RewriteTest {
             .activateRecipes("test.recipe.a")),
           text("Hi", "NoArgRecipeHi"));
     }
+
+    @Test
+    void declarativeRecipeChainFromResources() {
+        rewriteRun(spec -> spec.recipeFromResources("test.declarative.sample.a"),
+          text("Hi", "after"));
+    }
+
+    @Test
+    void declarativeRecipeChainFromResourcesIncludesImperativeRecipesInDescriptors() {
+        rewriteRun(spec -> spec.recipeFromResources("test.declarative.sample.a")
+            .afterRecipe(recipeRun -> assertThat(recipeRun.getChangeset().getAllResults().get(0)
+              .getRecipeDescriptorsThatMadeChanges().get(0).getRecipeList().get(0).getRecipeList().get(0)
+              .getDisplayName()).isEqualTo("Change text")),
+          text("Hi", "after"));
+    }
+
+    @Test
+    void declarativeRecipeChainFromResourcesLongList() {
+        rewriteRun(spec -> spec.recipe(Environment.builder()
+            .load(new YamlResourceLoader(RecipeLifecycleTest.class.getResourceAsStream("/META-INF/rewrite/test-sample-a.yml"), URI.create("rewrite.yml"), new Properties()))
+            .load(new YamlResourceLoader(RecipeLifecycleTest.class.getResourceAsStream("/META-INF/rewrite/test-sample-b.yml"), URI.create("rewrite.yml"), new Properties()))
+            .build()
+            .activateRecipes("test.declarative.sample.a")),
+          text("Hi", "after"));
+    }
 }
 
 class DefaultConstructorRecipe extends Recipe {

--- a/rewrite-core/src/test/resources/META-INF/rewrite/test-sample-a.yml
+++ b/rewrite-core/src/test/resources/META-INF/rewrite/test-sample-a.yml
@@ -1,3 +1,19 @@
+#
+# Copyright 2023 the original author or authors.
+# <p>
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+# <p>
+# https://www.apache.org/licenses/LICENSE-2.0
+# <p>
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
 ---
 type: specs.openrewrite.org/v1beta/recipe
 name: test.declarative.sample.a

--- a/rewrite-core/src/test/resources/META-INF/rewrite/test-sample-a.yml
+++ b/rewrite-core/src/test/resources/META-INF/rewrite/test-sample-a.yml
@@ -1,0 +1,6 @@
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: test.declarative.sample.a
+displayName: Test Recipe
+recipeList:
+  - test.declarative.sample.b

--- a/rewrite-core/src/test/resources/META-INF/rewrite/test-sample-b.yml
+++ b/rewrite-core/src/test/resources/META-INF/rewrite/test-sample-b.yml
@@ -1,0 +1,7 @@
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: test.declarative.sample.b
+displayName: Test Recipe
+recipeList:
+  - org.openrewrite.text.ChangeText:
+      toText: after

--- a/rewrite-core/src/test/resources/META-INF/rewrite/test-sample-b.yml
+++ b/rewrite-core/src/test/resources/META-INF/rewrite/test-sample-b.yml
@@ -1,3 +1,19 @@
+#
+# Copyright 2023 the original author or authors.
+# <p>
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+# <p>
+# https://www.apache.org/licenses/LICENSE-2.0
+# <p>
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
 ---
 type: specs.openrewrite.org/v1beta/recipe
 name: test.declarative.sample.b

--- a/rewrite-gradle/src/main/java/org/openrewrite/gradle/plugins/AddSettingsPluginRepository.java
+++ b/rewrite-gradle/src/main/java/org/openrewrite/gradle/plugins/AddSettingsPluginRepository.java
@@ -47,6 +47,7 @@ public class AddSettingsPluginRepository extends Recipe {
 
     @Option(displayName = "URL",
             description = "The url of the artifact repository",
+            required = false,
             example = "https://repo.spring.io")
     String url;
 

--- a/rewrite-test/src/main/java/org/openrewrite/test/RecipeSpec.java
+++ b/rewrite-test/src/main/java/org/openrewrite/test/RecipeSpec.java
@@ -127,6 +127,13 @@ public class RecipeSpec {
         return recipe(Objects.requireNonNull(RecipeSpec.class.getResourceAsStream(yamlResource)), activeRecipes);
     }
 
+    public RecipeSpec recipeFromResources(String... activeRecipes) {
+        return recipe(Environment.builder()
+                .scanYamlResources()
+                .build()
+                .activateRecipes(activeRecipes));
+    }
+
     /**
      * @param parser The parser supplier to use when a matching source file is found.
      * @return The current recipe spec.

--- a/rewrite-test/src/main/java/org/openrewrite/test/RecipeSpec.java
+++ b/rewrite-test/src/main/java/org/openrewrite/test/RecipeSpec.java
@@ -19,13 +19,11 @@ import com.fasterxml.jackson.databind.MapperFeature;
 import com.fasterxml.jackson.dataformat.csv.CsvMapper;
 import com.fasterxml.jackson.dataformat.csv.CsvSchema;
 import lombok.Getter;
-import lombok.RequiredArgsConstructor;
 import org.intellij.lang.annotations.Language;
 import org.openrewrite.*;
 import org.openrewrite.config.CompositeRecipe;
 import org.openrewrite.config.Environment;
 import org.openrewrite.config.YamlResourceLoader;
-import org.openrewrite.internal.InMemoryLargeSourceSet;
 import org.openrewrite.internal.lang.Nullable;
 
 import java.io.ByteArrayInputStream;
@@ -116,7 +114,6 @@ public class RecipeSpec {
 
     public RecipeSpec recipe(InputStream yaml, String... activeRecipes) {
         return recipe(Environment.builder()
-                .scanRuntimeClasspath() // Slow but required to find recipe classes the yaml recipe may depend on
                 .load(new YamlResourceLoader(yaml, URI.create("rewrite.yml"), new Properties()))
                 .build()
                 .activateRecipes(activeRecipes));

